docs/architecture/adr-033-protobuf-inter-module-comm.md

```diff
@@ -83,45 +83,64 @@ and `MsgClient` interfaces and propose this mechanism as a replacement for the e
 transaction
 
 This mechanism has the added benefits of:
 - reducing boilerplate through code generation, and
-- allowing for modules in other languages either via a VM like CosmWasm or sub-processes using gRPC 
+- allowing for modules in other languages either via a VM like CosmWasm or sub-processes using gRPC
 
 ### Inter-module Communication
 
-To use the `Client` generated by the protobuf compiler we need a `grpc.ClientConn` implementation. For this we introduce
+To use the `Client` generated by the protobuf compiler we need a `grpc.ClientConn` [interface](https://github.com/regen-network/protobuf/blob/cosmos/grpc/types.go#L12)
+implementation. For this we introduce
 a new type, `ModuleKey`, which implements the `grpc.ClientConn` interface. `ModuleKey` can be thought of as the "private
-key" corresponding to a module account, where authentication is provided through use of a special `Invoker()` function, 
+key" corresponding to a module account, where authentication is provided through use of a special `Invoker()` function (gRPC middleware),
 described in more detail below.
 
-Whereas external clients use their account's private key to sign transactions containing `Msg`s where they are listed as signers,
-modules use their `ModuleKey` to send `Msg`s where they are listed as the sole signer to other modules. For example, modules
-could use their `ModuleKey` to "sign" a `/cosmos.bank.Msg/Send` transaction to send coins from the module's account to another
-account.
+Blockchain users (external clients) use their account's private key to sign transactions containing `Msg`s where they are listed as signers (each
+message specifies required signers with `Msg.GetSigner`). The authentication checks is performed by `AnteHandler`.
+
+Here, we extend this process, by allowing modules to be identified in `Msg.GetSigners`. When a module creates a `Msg` for another module,
+it's `ModuleKey` is used to send (through the `ClientConn` interface we describe below) and is set as a sole "signer". It's worth to note
+that we don't store any cryptography signature in this case.
+For example, module `A` could use it's `A.ModuleKey` to create `MsgSend` object for `/cosmos.bank.Msg/Send` transaction. `MsgSend` validation
+will assure that the `from` account (`A.ModuleKey` in this case) is signer.
 
 Here's an example of a hypothetical module `foo` interacting with `x/bank`:
 ```go
 package foo
 
-func (fooMsgServer *MsgServer) Bar(ctx context.Context, req *MsgBarRequest) (*MsgBarResponse, error) {
-  bankQueryClient := bank.NewQueryClient(fooMsgServer.moduleKey)
-  balance, err := bankQueryClient.Balance(&bank.QueryBalanceRequest{Address: fooMsgServer.moduleKey.Address(), Denom: "foo"})
-  
+
+type FooMsgServer {
+  // ...
+
+  bankQuery bank.QueryClient
+  bankMsg   bank.MsgClient
+}
+
+func NewFooMsgServer(...) MsgServer {
```

Comment:
`moduleKey` should probably be a constructor param

Reply:
Yes, good idea :+1: